### ISSUE TITLE
Splicing an array with an expression for the boundary indices fails

### ISSUE THREAD

**Issue by Kopfbremse** [mikefarah/yq#1419]
**Version of yq**: 4.29.2
**Operating system**: linux
**Installed via**: binary release

**Describe the bug**
I want to insert an element into an array after an existing element with a certain property.
This fails when the boundary indices are calculated with an expression.

**Input Yaml**
```yaml
names:
  - name: first
  - name: last
```

-----------
**Code to find the correct position**
```bash
yq e '.names | to_entries | .[] | select(.value.name == "first") | .key + 1'
```

Output
```yaml
1
```

-----------
**Code to insert an element into an array**
```bash
yq e '.names |= .[0:1] +
                [{"name": "middle"}] +
                .[1:]'
```

Output
```yaml
names:
  - name: first
  - name: middle
  - name: last
```

-----------
**Combining both steps, error 1**
```bash
yq e '
  .names |= .[0: (to_entries | .[] | select(.value.name == "first" | .key + 1))] +
            [{"name": "middle"}] +
            .[1:]'
```

Output
```text
Error: cannot index array with '' (strconv.ParseInt: parsing "": invalid syntax)
```

-----------
**Combining both steps, error 2**
```bash
yq e '
  .names |= .[0:1] +
            [{"name": "middle"}] +
            .[(to_entries | .[] | select(.value.name == "first" | .key + 1)):]'
```

Output
```text
Error: '|=' expects 2 args but there is 1
```

**Comment by mikefarah** [comment]
Oh I hadn't considered expressions for splice - will fix.

**Comment by Kopfbremse** [comment]
In this use case I need the expression `(to_entries | .[] | select(.value.name == "first" | .key + 1))` twice.

**Command**
```bash
yq e '
  .names |= .[0: (to_entries | .[] | select(.value.name == "first" | .key + 1))] +
            [{"name": "middle"}] +
            .[(to_entries | .[] | select(.value.name == "first" | .key + 1)):]'
```

Will it ever be possible to define your own functions in yq as provided by [jq](https://stedolan.github.io/jq/) to follow the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) encoding rule?
That would be great!

**Comment by mikefarah** [comment]
Fixed in 4.30.1 - that would definitely be cool, however in this case you can use a variable to DRY the expression, see https://mikefarah.gitbook.io/yq/operators/slice-array#inserting-into-the-middle-of-an-array
